quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
         # Svn doesn't support local sandbox ignore lists
         pass
 
+class GitlabSubproject(Subproject):
+    def __init__(self, name, url, directory, options, conf={}, **kwargs):
+        super().__init__(name, directory, options, conf, **kwargs)
+
+        self.stamp = {
+            "job_id": None,  # GitLab job ID. Mostly used for freezing gitlab+ci URLs.
+            "sha1": None,  # SHA1 of downloaded artifact. Mostly used for freezing.
+            "url": urllib.parse.urlunparse(url),  # type: ignore
+        }  # type: dict[str, None | str]
+
+        self._gitlab_setup(url)
+        self._parse_url(url)
+
+    def update(self, clean=False, fix_remotes=False):
+        assert self.directory  # Make Pyright happy
+
+        # Check whether the last download matches the currently-wanted one.
+        #
+        # NOTE: The current logic will force a re-download after a freeze (the URL changes).
+        stamp_file = join(self.directory, ".stamp.quark")
+        if exists(join(stamp_file)):
+            with open(stamp_file, "r") as fileobj:
+                data = json.load(fileobj)
+                if data["job_id"] == self.stamp["job_id"] and data["url"] == self.stamp["url"]:
+                    return
+
+        # The stamp has changed and the directory already exists: we need to start from scratch.
+        if isdir(self.directory):
+            shutil.rmtree(self.directory)
+
+        with tempfile.TemporaryDirectory(prefix="quark-") as tempdir:
+            archive_path = self._download(tempdir)
+
+            if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz", ".zip")):
```

Comment:
Deciding to extract depending on the extension of the downloaded file doesn't seem like a good idea... if the Gitlab APIs always give us—say—a .zip containing the actual artifact, then we should always extract; if instead we always get the actual file the user requested, it should be left alone: the downloaded artifact itself should be an opaque to this code, and it's a legitimate use case to download an artifact that is a zip file that should be left alone (we have several such cases for camera/beaglebone firmware stuff, that are .tar.gz shipped as is in other software).